### PR TITLE
Query block

### PR DESCRIPTION
**Pull request by iruzevic**
# Worklog
- Added new block called query block
- You can add data using the filter from your project
- This block can be used to add any kind of data in the user selection that is project specific

# Screens

https://user-images.githubusercontent.com/23283324/141752547-041352b6-248a-453e-ab27-31416a81cbbd.mov

### CODE REVIEW

File: src/Blocks/custom/query/components/query-editor.js [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+	const unique = useMemo(() => getUnique(), []);
+
+	const {
+		blockClass,
+		blockFullName
+	} = attributes;
+
+	const formPostId = useSelect((select) => select('core/editor').getCurrentPostId());
+
+		// Populate ID manually and make it generic.
```
Comment (goranalkovic-infinum) [inline src/Blocks/custom/query/components/query-editor.js:17]: Please check indent here

File: src/Hooks/Filters.md [hunk 2]
```diff
@@ -158,3 +158,90 @@ public function getFieldStyleOptions(): array
 	];
 }
 ```
+
+## Add data to query block
+
+This two filters will add necessary data to the query block. This block can show any kind of data provided via filter.
```
Comment (goranalkovic-infinum) [inline src/Hooks/Filters.md:164]:
```suggestion
These filters will add the necessary data for the Query block to work.
```

File: src/Hooks/Filters.md [hunk 3]
```diff
@@ -158,3 +158,90 @@ public function getFieldStyleOptions(): array
...
+
+Field data option selector will not show unless a filter is provided.
```
Comment (goranalkovic-infinum) [inline src/Hooks/Filters.md:166]:
```suggestion
Field data option selector will not be shown unless a filter is added.
```

File: src/Hooks/Filters.md [hunk 4]
```diff
@@ -158,3 +158,90 @@ public function getFieldStyleOptions(): array
...
+
+**Default values:**
+```php
+[]
+```
+
+**Filter for providing editor options:**
```
Comment (goranalkovic-infinum) [inline src/Hooks/Filters.md:173]:
```suggestion
**Filter for providing Block editor options:**
```

File: src/Hooks/Filters.md [hunk 5]
```diff
@@ -158,3 +158,90 @@ public function getFieldStyleOptions(): array
...
+```php
+// Provide custom query block options.
+add_filter('es_forms_block_query_options', [$this, 'getQueryBlockOptions']);
+
+/**
+ * Provide custom query block options.
```
Comment (goranalkovic-infinum) [inline src/Hooks/Filters.md:179]:
```suggestion
 * Provide custom Query block options.
```

File: src/Hooks/Filters.md [hunk 6]
```diff
@@ -158,3 +158,90 @@ public function getFieldStyleOptions(): array
...
+			"label" => "Jobs",
+			"value" => "jobs"
+		],
+	];
+}
+```
+
+**Filter for providing options data:**
+```php
+// Provide custom query block options.
```
Comment (goranalkovic-infinum) [inline src/Hooks/Filters.md:200]:
```suggestion
// Provide custom Query block options.
```

File: src/Hooks/Filters.md [hunk 7]
```diff
@@ -158,3 +158,90 @@ public function getFieldStyleOptions(): array
...
+		],
+		[
+			"label" => "Jobs",
+			"value" => "jobs"
+		],
+	];
+}
+```
+
+**Filter for providing options data:**
```
Comment (goranalkovic-infinum) [inline src/Hooks/Filters.md:198]:
```suggestion
**Filter for providing option data:**
```

File: src/Hooks/Filters.md [hunk 8]
```diff
@@ -158,3 +158,90 @@ public function getFieldStyleOptions(): array
...
+```php
+// Provide custom query block options.
+add_filter('es_forms_block_query_options_data', [$this, 'getQueryBlockOptionsData']);
+
+/**
+ * Provide custom query block options data.
```
Comment (goranalkovic-infinum) [inline src/Hooks/Filters.md:204]:
```suggestion
 * Provide custom Query block options data.
```

File: src/Hooks/Filters.md [hunk 9]
```diff
@@ -158,3 +158,90 @@ public function getFieldStyleOptions(): array
...
+ *
+ * @param string $type Type of option selected in the editor.
```
Comment (goranalkovic-infinum) [inline src/Hooks/Filters.md:206]:
```suggestion
 * @param string $type Type of option selected in the Block editor.
```

**Review by goranalkovic-infinum — CHANGES_REQUESTED** [review]
Awesome job! 🎉 

Left a couple of suggestions (mostly grammar & typos), but there's one bigger thing:

**We should consider naming this block something else**, as Core already has a _Query_ block (used as a front for WP_Query) and besides taking the filter values, I don't see too much querying going on here 😄 

Maybe _Custom data_ or _Selectable data_ would be a better name?

**Review by iobrado — APPROVED** [review]

**Comment by iruzevic** [comment]
> Awesome job! 🎉
> 
> Left a couple of suggestions (mostly grammar & typos), but there's one bigger thing:
> 
> **We should consider naming this block something else**, as Core already has a _Query_ block (used as a front for WP_Query) and besides taking the filter values, I don't see too much querying going on here 😄
> 
> Maybe _Custom data_ or _Selectable data_ would be a better name?

I have refactored this to custom-data block

File: src/Blocks/custom/custom-data/manifest.json [hunk 10]
```diff
@@ -1,30 +1,30 @@
 {
-	"blockName": "query",
-	"title": "Query",
-	"description" : "Block used to show custom query build using WordPress filters.",
+	"blockName": "custom-data",
+	"title": "customData",
```
Comment (goranalkovic-infinum) [inline src/Blocks/custom/custom-data/manifest.json:3]:
```suggestion
	"title": "Custom data",
```

File: src/Blocks/custom/custom-data/custom-data.php [hunk 11]
```diff
@@ -0,0 +1,34 @@
...
+
+$customDataClass = Components::classnames([
+	Components::selector($blockClass, $blockClass),
+	Components::selector(!$block, $blockClass, '', 'invalid')
+]);
+
+if (!$block) {
+	?>
+	<div class="<?php echo esc_attr($customDataClass); ?>">
+		<?php esc_html_e('Sorry, it looks like your custom data block not configured correctly. In order for this block to work you must provide data using our filters. Please check our documentation for details.', 'eightshift-forms'); ?>
```
Comment (goranalkovic-infinum) [inline src/Blocks/custom/custom-data/custom-data.php:29]:
```suggestion
		<?php esc_html_e('Sorry, it looks like your Custom data block is not configured correctly. In order for the block to work, you must provide data using filters. Check the documentation for details.', 'eightshift-forms'); ?>

```

**Review by goranalkovic-infinum — APPROVED** [review]
Couple more suggestions 😄

File: src/Blocks/custom/custom-data/custom-data-editor.scss [hunk 12]
```diff
@@ -0,0 +1,20 @@
+.es-block-custom-data {
+	// Transparent overlay over the forms so the user can't click on them from the admin screen.
+	&::after {
+		content: '';
+		background-color: transparent;
+		width: 100%;
+		height: 100%;
+		position: absolute;
+		left: 0;
+		top: 0;
```
Comment (volfkarlo) [inline src/Blocks/custom/custom-data/custom-data-editor.scss:10]: What you can add to reassure the non-clicking ability is 
`pointer-events: all;` this should ensure to receive all of the events over the after component.

**Review by volfkarlo — APPROVED** [review]
👏  job well done, just a minor potential improvement.

I'll approve it because it is probably not necessary but more of a reassurance.

**Review by mbmjertan — APPROVED** [review]

File: src/Blocks/custom/custom-data/custom-data-editor.scss (hunk 12 above)
Comment (iruzevic) [reply src/Blocks/custom/custom-data/custom-data-editor.scss:10]: I will implement this thx

File: src/Blocks/BlockCustomData.php [hunk 13]
```diff
@@ -0,0 +1,125 @@
...
+					'checkboxesName' => $attributes['customDataCheckboxesName'] ?? '',
+					'checkboxesContent' => array_map(
+						function ($option) {
+							return [
+								'component' => 'checkbox',
+								'checkboxLabel' => $option['label'] ?? '',
+								'checkboxValue' => $option['value'] ?? $option['label'] ?? '',
+								'checkboxIsChecked' => $option['selected'] ?? false,
+							];
+						},
```
Comment (dingo-d) [inline src/Blocks/BlockCustomData.php:78]: I haven't got the time to comment, but you can open a new PR just for this.

Wherever you are using anonymous functions (like here in `array_map`) if you are not using `$this` keyword in a function (callback) you can improve the performance by adding a `static` in front of the `function`.

Like:

```php
'checkboxesContent' => array_map(
	static function ($option) {
		return [
			'component' => 'checkbox',
			'checkboxLabel' => $option['label'] ?? '',
			'checkboxValue' => $option['value'] ?? $option['label'] ?? '',
			'checkboxIsChecked' => $option['selected'] ?? false,
		];
	},
	$data
),
```

Comment (iruzevic) [reply src/Blocks/BlockCustomData.php:78]: awesome did't know that. I have opened a issue tnx

**Pull request merged**